src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class FileLoggerProcessor : IDisposable
+    {
+        private const int _maxQueuedMessages = 1024;
+
+        private readonly string _path;
+        private readonly string _fileName;
+        private readonly int? _maxFileSize;
+        private readonly int? _maxRetainedFiles;
+        private int _fileNumber = 1;
+
+        private readonly BlockingCollection<LogMessage> _messageQueue = new BlockingCollection<LogMessage>(_maxQueuedMessages);
+        private readonly List<LogMessage> _currentBatch = new List<LogMessage>();
+        private Task _outputTask;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        public FileLoggerProcessor(IOptionsMonitor<FileLoggerOptions> options)
+        {
+            var loggerOptions = options.CurrentValue;
+            _path = loggerOptions.LogDirectory;
+            _fileName = loggerOptions.FileName;
+            _maxFileSize = loggerOptions.FileSizeLimit;
+            _maxRetainedFiles = loggerOptions.RetainedFileCountLimit;
+
+            // Start message queue processor
+            _cancellationTokenSource = new CancellationTokenSource();
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public void EnqueueMessage(LogMessage message)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                while (_messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                }
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch);
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+
+                    _currentBatch.Clear();
+                }
+            }
+        }
+
+        internal async Task WriteMessagesAsync(IEnumerable<LogMessage> messages)
```

Comment:
No need to use interface types or enumerables on internal APIs (why not private?) if you don't have to